staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,612 @@
+		// that virtual time advanced due to the requests actually executing.
+
+		// reqs += len(queue.Requests)
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
+		return 0
+	}
+	return math.Min(float64(reqs), float64(qs.concurrencyLimit)) / float64(NEQ)
+}
+
+// updateQueueVirStartTime updates the virtual start time for a queue
+// this is done when a new packet is enqueued.  For more info see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching
+func (qs *queueSetImpl) updateQueueVirStartTime(packet *Request, queue *Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.Requests) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.Requests) == 1 && queue.RequestsExecuting == 0 {
+		// the queue’s virtual start time is set to the virtual time.
+		queue.VirStart = qs.vt
+	}
+}
+
+// removeQueueAndUpdateIndexes uses reslicing to remove an index from a slice
+//  and then updates the 'Index' field of the queues to be correct
+func removeQueueAndUpdateIndexes(queues []*Queue, index int) []*Queue {
+	removedQueues := removeIndex(queues, index)
+	for i := index; i < len(removedQueues); i++ {
+		removedQueues[i].Index--
+	}
+	return removedQueues
+}
+
+// removeIndex uses reslicing to remove an index from a slice
+func removeIndex(s []*Queue, index int) []*Queue {
+	return append(s[:index], s[index+1:]...)
+}
+
+// FinishRequestAndDequeueWithChannelAsMuchAsPossible is a convenience method which calls finishRequest
+// for a given packet and then dequeues as many packets as possible
+// and updates that packet's channel signifying it is is dequeued
+// this is a callback used for the filter that the queueSetImpl supports
+func (qs *queueSetImpl) FinishRequestAndDequeueWithChannelAsMuchAsPossible(pkt *Request) {
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
+
+	qs.finishRequest(pkt)
+	qs.DequeueWithChannelAsMuchAsPossible()
+}
+
+// FinishRequest is a callback that should be used when a previously dequeued packet
+// has completed it's service.  This callback updates imporatnt state in the
+// queueSetImpl
+func (qs *queueSetImpl) finishRequest(p *Request) {
+
+	S := qs.clk.Since(p.StartTime).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	p.Queue.VirStart -= qs.estimatedServiceTime - S
+
+	// request has finished, remove from requests executing
+	p.Queue.RequestsExecuting--
+
+	// Logic to remove quiesced queues
+	// >= as QueueIdx=25 is out of bounds for desiredNumQueues=25 [0...24]
+	if p.Queue.Index >= qs.desiredNumQueues &&
+		len(p.Queue.Requests) == 0 &&
+		p.Queue.RequestsExecuting == 0 {
+		qs.queues = removeQueueAndUpdateIndexes(qs.queues, p.Queue.Index)
```

Comment:
At this point, to be precise, `qs.robinIdx` should be decremented if it is `>= p.Queue.Index` because we want to maintain the invariant that `(qs.robinIdx+1) % numQueues` is the index of the next queue after the one last dispatched from.

However, reconfiguring should be rare, so a little skip in the round-robin at this time could be forgiven.

Reply:
Done